cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/queue/hbase/HBaseConsumerStateStore.java

```diff
@@ -15,44 +15,260 @@
-    hTable.flushCommits();
+  /**
+   * Updates the start row state of the given consumer.
+   */
+  void updateState(long groupId, int instanceId, byte[] startRow) {
+    byte[] stateColumn = getConsumerStateColumn(groupId, instanceId);
+    table.put(queueName.toBytes(), stateColumn, startRow);
+  }
+
+  /**
+   * Remove all states related to the queue that this state store is representing.
+   */
+  void clear() {
+    // Scan and delete all rows prefix with queueName
+    Scanner scanner = table.scan(queueName.toBytes(), Bytes.stopKeyForPrefix(queueName.toBytes()));
+    Row row = scanner.next();
+    while (row != null) {
+      table.delete(row.getRow());
+      row = scanner.next();
+    }
+  }
+
+  /**
+   * Changes the number of consumer instances of the given consumer group. The consumer group configuration needs to be
+   * existed already.
+   *
+   * @param groupId groupId of the consumer group
+   * @param instances number of instances to change to
+   */
+  @Override
+  public void configureInstances(long groupId, int instances) {
+    // Get all latest entry row key of all existing instances
+    Map<Integer, byte[]> startRows = fetchStartRows(groupId);
+    int oldInstances = startRows.size();
+
+    // Nothing to do if size doesn't change
+    if (oldInstances == instances) {
+      return;
+    }
+    // Compute and applies changes
+    updateGroupInstances(groupId, instances, startRows);
+  }
+
+  /**
+   * Changes the configuration of all consumer groups. For groups with existing states but not in the given group info,
+   * their states will be removed.
+   *
+   * @param groupInfo map from groupId to number of instances for that group
+   */
+  @Override
+  public void configureGroups(Map<Long, Integer> groupInfo) {
+    byte[] rowKey = queueName.toBytes();
+    Row row = table.get(rowKey);
+
+    // Generate existing groupInfo, also find smallest rowKey from existing group if there is any
+    Map<byte[], byte[]> columns = row.getColumns();
+    if (columns == null) {
+      columns = ImmutableSortedMap.of();
+    }
+    Map<Long, Integer> oldGroupInfo = Maps.newHashMap();
+    byte[] smallest = decodeGroupInfo(groupInfo, columns, oldGroupInfo);
+
+    // For groups that are removed, simply delete the columns
+    for (long removeGroupId : Sets.difference(oldGroupInfo.keySet(), groupInfo.keySet())) {
+      for (int instanceId = 0; instanceId < oldGroupInfo.get(removeGroupId); instanceId++) {
+        table.delete(rowKey, getConsumerStateColumn(removeGroupId, instanceId));
+      }
+    }
+
+    // For each group that changed (either a new group or number of instances change), update the startRow
+    for (Map.Entry<Long, Integer> entry : groupInfo.entrySet()) {
+      long groupId = entry.getKey();
+      int instances = entry.getValue();
+      if (!oldGroupInfo.containsKey(groupId)) {
+        // For new group, simply put with smallest rowKey from other group or an empty byte array if none exists.
```

Comment:
So this happens when a flow is updated and a new flowlet is added? Or how else can that happen?
In that case, where should the new flowlet start reading? Perhaps the current time? 
It seems arbitrary to just use some other group's minStartRow...